### PR TITLE
Feat/connections search function

### PR DESCRIPTION
**Pull request by Ellenn-A**
# Pull Request

## Checklist
- [x] Have you read Digital Catapult's [Code of Conduct](https://github.com/digicatapult/.github/blob/main/CODE_OF_CONDUCT.md)?
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

## PR Type

Please delete options that are irrelevant.

- [ ] Bug Fix
- [x] Feature
- [ ] Documentation Update
- [ ] Code style update (formatting, local variables)
- [ ] Breaking Change (fix or feature that would cause existing functionality to change)

## Linked tickets

- https://digicatapult.atlassian.net/browse/VR-65

## High level description

Search functionality within the list of existing connections.

## Detailed description

Search through connections based on company name. The minimum to search is 3 characters in the search window. If there are no connections with matching company names - show a message. 
Screenshot 2024-06-05 at 14 09 10
Screenshot 2024-06-05 at 14 09 27

Screenshot 2024-06-05 at 14 09 51

## Describe alternatives you've considered

N/A

## Operational impact

Will be able to search through a list of connections. 

## Additional context

Just an fyi I have increased the seed file to test this properly, we might want to consider keeping original seed file. Please let me know your thoughts.

### CODE REVIEW

File: seeds/connections-seed.js [hunk 1]
```diff
@@ -1,17 +1,78 @@
...
+//   { status: 'unverified', company_number: company_number + '3', company_name },
+//   { status: 'verified_them', company_number: company_number + '4', company_name },
+//   { status: 'verified_us', company_number: company_number + '5', company_name },
+//   { status: 'verified_both', company_number: company_number + '6', company_name },
+//   { status: 'disconnected', company_number: company_number + '7', company_name },
+// ]
+
+export async function prepareVariants() {
+  const variants10000 = []
+  for (let i = 0; i < 10000; i++) {
```
Comment (mattdean-digicatapult) [inline seeds/connections-seed.js:19]: can we make the `10000` a const at the top of the file and then make you branches conditional on `i % 7`. That way we can adjust the import size more easily

File: src/controllers/connection/index.ts [hunk 2]
```diff
@@ -26,9 +26,12 @@ export class ConnectionController extends HTMLController {
...
   @SuccessResponse(200)
   @Get('/')
-  public async listConnections(): Promise<HTML> {
+  public async listConnections(@Query() search?: string): Promise<HTML> {
     this.logger.debug('connections page requested')
-    const connections = await this.db.get('connection', {}, [['updated_at', 'desc']])
-    return this.html(this.connectionTemplates.listPage(connections))
+    const query = search ? [['company_name', 'ILIKE', `%${search}%`]] : {}
+    const connections = await this.db.get('connection', query, [['updated_at', 'desc']])
+    console.log(connections)
```
Comment (mattdean-digicatapult) [inline src/controllers/connection/index.ts:33]: remove console log

File: src/controllers/connection/index.ts [hunk 3]
```diff
@@ -26,9 +26,12 @@ export class ConnectionController extends HTMLController {
    */
   @SuccessResponse(200)
   @Get('/')
-  public async listConnections(): Promise<HTML> {
+  public async listConnections(@Query() search?: string): Promise<HTML> {
```
Comment (mattdean-digicatapult) [inline src/controllers/connection/index.ts:29]: need to update unit tests on controller

File: src/models/db/types.ts [hunk 4]
```diff
@@ -60,7 +60,7 @@ export type Models = {
 type WhereComparison<M extends TABLE> = {
   [key in keyof Models[M]['get']]: [
     Extract<key, string>,
-    '=' | '>' | '>=' | '<' | '<=' | '<>',
+    '=' | '>' | '>=' | '<' | '<=' | '<>' | 'LIKE' | 'ILIKE' | '~',
```
Comment (mattdean-digicatapult) [inline src/models/db/types.ts:63]: let's remove `~` as we're not using it or planning on. I'm not sure the impact regular expresssions would have on things

File: public/styles/main.css [hunk 5]
```diff
@@ -252,6 +252,9 @@ a.connections-table.icon {
     justify-content: space-between;
     min-height: auto;
   }
+  .connections.list.nav {
```
Comment (mattdean-digicatapult) [inline public/styles/main.css:255]: can we hyphenate this please instead o fmultiple classes?

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]

File: public/styles/main.css (hunk 5 above)
Comment (Ellenn-A) [reply public/styles/main.css]: done and I hyphenated it in the other place it was used in too

File: seeds/connections-seed.js [hunk 6]
```diff
@@ -1,17 +1,80 @@
...
+// const variants = [
+//   { status: 'pending', company_number: company_number + '0', company_name },
+//   { status: 'pending', company_number: company_number + '1', company_name: '_@@$@*@&%£*&@£*$*(@£*$)@£*&' },
+//   { status: 'unverified', company_number: company_number + '2', company_name: 0 },
+//   { status: 'unverified', company_number: company_number + '3', company_name },
+//   { status: 'verified_them', company_number: company_number + '4', company_name },
+//   { status: 'verified_us', company_number: company_number + '5', company_name },
+//   { status: 'verified_both', company_number: company_number + '6', company_name },
+//   { status: 'disconnected', company_number: company_number + '7', company_name },
+// ]
```
Comment (mattdean-digicatapult) [inline seeds/connections-seed.js:15]: let's remove the old seed

File: seeds/connections-seed.js [hunk 7]
```diff
@@ -1,17 +1,80 @@
...
+
+const variantsize = 10000
+
+export async function prepareVariants() {
+  const variants10000 = []
+  for (let i = 0; i < variantsize; i++) {
+    if (i % 7 == 0) {
+      variants10000.push({
+        status: 'pending',
+        company_number: i.toString(),
```
Comment (mattdean-digicatapult) [inline seeds/connections-seed.js:25]: can we pad the numbers to make them valid. You can just do `i.toString().padStart(8, '0')`

File: seeds/connections-seed.js [hunk 8]
```diff
@@ -1,17 +1,80 @@
...
+      variants10000.push({
+        status: 'disconnected',
+        company_number: i.toString(),
+        company_name: `Some other company_name Name ${i.toString()}`,
+      })
+    }
+  }
+  variants10000.push({
+    status: 'disconnected',
+    company_number: '009',
```
Comment (mattdean-digicatapult) [inline seeds/connections-seed.js:68]: you may want to make these speciail ones start with a 1 such as `10000009` to avoid clashes with the loop

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
Just a few ocmments on the seed

**Review by mattdean-digicatapult — APPROVED** [review]

**Pull request merged**